### PR TITLE
[rb] Add Bidi network commands for authentication and interception

### PR DESCRIPTION
**Pull request by aguspe**
### Description
This PR adds the following:

- A Network bidi class that allows to add, remove and clear auth handlers
- A Network class to be call from driver
- Unit and integration tests

### Motivation and Context
Based on #13993 this PR follows the agreed defined API methods implemented for the Ruby binding, now is possible to call driver.network.add_authentication_handler

### Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Checklist
- [x] I have read the [contributing](https://github.com/SeleniumHQ/selenium/blob/trunk/CONTRIBUTING.md) document.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

___

### **PR Type**
Enhancement, Tests

___

### **Description**
- Introduced a new `Network` class within the `BiDi` module to handle network events and authentication.
- Added methods for intercepting network requests and managing authentication handlers.
- Implemented integration and unit tests for the new `Network` class to ensure functionality.
- Updated type signatures for the `Network` class and related methods.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><details><summary>5 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>bidi.rb</strong><dd><code>Add autoload for Network in BiDi class</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

rb/lib/selenium/webdriver/bidi.rb

- Added autoload for `Network` in the `BiDi` class.

</details>

  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/14523/files#diff-b7660c7b53fec20ce95452b2afdbf306e9c36363e5bf8812863375b7bc96e42e">+1/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>network.rb</strong><dd><code>Introduce Network class with intercept and auth handling</code>&nbsp; </dd></summary>
<hr>

rb/lib/selenium/webdriver/bidi/network.rb

<li>Introduced a new <code>Network</code> class within <code>BiDi</code>.<br> <li> Defined network events and phases.<br> <li> Implemented methods for intercepting network requests and handling <br>authentication.<br>

</details>

  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/14523/files#diff-caed96b819b484a7727db49b5712d0caa86ddf86807a544fc8bd0a94756e301c">+70/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>common.rb</strong><dd><code>Add require statement for network</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

rb/lib/selenium/webdriver/common.rb

- Added require statement for `network`.

</details>

  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/14523/files#diff-5bf2b6715d2e2a368c57757a030c83091c6d52e4db8175f66df8b98b12ac0d7a">+1/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>driver.rb</strong><dd><code>Add network method to Driver class</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

rb/lib/selenium/webdriver/common/driver.rb

- Added `network` method to return a `Network` instance.

</details>

  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/14523/files#diff-ecf89969518aec5e73ce3aff22dc10e88185534845a8a17228952c92f1498cdd">+9/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>network.rb</strong><dd><code>Create Network class for authentication management</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

rb/lib/selenium/webdriver/common/network.rb

<li>Created a <code>Network</code> class for managing authentication handlers.<br> <li> Implemented methods to add and remove authentication handlers.<br>

</details>

  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/14523/files#diff-36bbaec2382eeda5ec71df015b4ed38c64e80feb08f44ba368349b07c434f044">+46/-0</a>&nbsp; &nbsp; </td>

</tr>                    
</table></details></td></tr><tr><td><strong>Formatting</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>struct.rb</strong><dd><code>Fix comment formatting in BiDi module</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

rb/lib/selenium/webdriver/bidi/struct.rb

- Fixed comment formatting for the `BiDi` module.

</details>

  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/14523/files#diff-04c05a628e46aa69c7d088546e665bcdd069b1869bc82d7e0466d4915624f3b7">+1/-3</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    
</table></details></td></tr><tr><td><strong>Tests</strong></td><td><details><summary>3 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>network_spec.rb</strong><dd><code>Add integration tests for BiDi Network class</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

rb/spec/integration/selenium/webdriver/bidi/network_spec.rb

<li>Added integration tests for <code>Network</code> class in <code>BiDi</code>.<br> <li> Tested intercept addition, removal, and authentication continuation.<br>

</details>

  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/14523/files#diff-e7c666a99d84b05ace437427175d9e64176f5628789114d383ed245836d3dd0d">+65/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>network_spec.rb</strong><dd><code>Add integration tests for Network class</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

rb/spec/integration/selenium/webdriver/network_spec.rb

<li>Added integration tests for <code>Network</code> class.<br> <li> Tested adding and removing authentication handlers.<br>

</details>

  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/14523/files#diff-1a5e512bf311a90fe92d313bab7fdf226fe29998c5b4bd92b78e354fa6948e36">+48/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>network_spec.rb</strong><dd><code>Add unit tests for BiDi Network class</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

rb/spec/unit/selenium/webdriver/bidi/network_spec.rb

<li>Added unit tests for <code>Network</code> class in <code>BiDi</code>.<br> <li> Mocked tests for intercept and authentication handling.<br>

</details>

  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/14523/files#diff-7871ccc27da6b58cdbf74858c3f84397416ecc66cfa294275d20a2ac0dace47a">+63/-0</a>&nbsp; &nbsp; </td>

</tr>                    
</table></details></td></tr><tr><td><strong>Documentation</strong></td><td><details><summary>3 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>network.rbs</strong><dd><code>Add type signatures for BiDi Network class</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

rb/sig/lib/selenium/webdriver/bidi/network.rbs

- Added type signatures for `Network` class in `BiDi`.

</details>

  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/14523/files#diff-ed9292ed35d71b668200e1c25116ba4a1933ec8df1aadd76ea9a14f7cc9a1763">+23/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>driver.rbs</strong><dd><code>Add type signature for network method</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

rb/sig/lib/selenium/webdriver/common/driver.rbs

- Added type signature for `network` method.

</details>

  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/14523/files#diff-91b084e4330efe2794c17d5b79d203958bd89469087f9455069256e9ec846bda">+2/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>network.rbs</strong><dd><code>Add type signatures for Network class</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

rb/sig/lib/selenium/webdriver/common/network.rbs

- Added type signatures for `Network` class.

</details>

  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/14523/files#diff-9e106b0b1db8e86ec1c738074989c75c8a3a0c60080d22d2d687eb0c5d42e89f">+15/-0</a>&nbsp; &nbsp; </td>

</tr>                    
</table></details></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**: Comment `/help "your question"` on any pull request to receive relevant information

### CODE REVIEW

File: rb/lib/selenium/webdriver/common/network.rb [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+  module WebDriver
+    class Network
+      attr_reader :auth_callbacks
+
+      def initialize(bridge)
+        @network = BiDi::Network.new(bridge.bidi)
+        @auth_callbacks = {}
+      end
+
+      def add_auth_handler(username, password)
```
Comment (p0deje) [inline rb/lib/selenium/webdriver/common/network.rb]:
```suggestion
      def add_authentication_handler(username, password)
```

File: rb/lib/selenium/webdriver/common/network.rb [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+        auth_id = @network.on(:auth_required) do |event|
+          request_id = event['requestId']
+          @network.continue_with_auth(request_id, username, password)
+        end
+        @auth_callbacks[auth_id] = intercept
+
+        auth_id
+      end
+
+      def remove_auth_handler(id)
```
Comment (p0deje) [inline rb/lib/selenium/webdriver/common/network.rb]:
```suggestion
      def remove_authentication_handler(id)
```

File: rb/lib/selenium/webdriver/common/network.rb [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+        intercept = @auth_callbacks[id]
+        @network.remove_intercept(intercept['intercept'])
+        @auth_callbacks.delete(id)
+      end
+
+      def clear_auth_handlers
```
Comment (p0deje) [inline rb/lib/selenium/webdriver/common/network.rb]:
```suggestion
      def clear_authentication_handlers
```

File: rb/spec/integration/selenium/webdriver/bidi/network_spec.rb [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+require_relative '../spec_helper'
+require_relative '../../../../../lib/selenium/webdriver/bidi/session'
```
Comment (p0deje) [inline rb/spec/integration/selenium/webdriver/bidi/network_spec.rb]: You should not need to load these files directly, they should be autoloaded by Selenium gem.

File: rb/spec/integration/selenium/webdriver/network_spec.rb [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+require_relative 'spec_helper'
+require_relative '../../../../lib/selenium/webdriver/common/network'
```
Comment (p0deje) [inline rb/spec/integration/selenium/webdriver/network_spec.rb]: Ditto.

File: rb/sig/lib/selenium/webdriver/common/driver.rbs [hunk 6]
```diff
@@ -17,6 +17,8 @@ module Selenium
 
       def inspect: () -> String
 
+      def network: -> untyped
```
Comment (p0deje) [inline rb/sig/lib/selenium/webdriver/common/driver.rbs]:
```suggestion
      def network: -> Network
```

**Comment by aguspe** [comment]
@p0deje whenever you have time I fixed the linter issues, so I can continue with the BiDi implementation for Ruby, and ran the linter locally to double check, thank you so much!

Screenshot 2024-11-09 at 16 47 19

**Pull request merged**